Feat/bamboo relay coordinator

Arctek opened the pull request:
Added support for the coordinator to the Bamboo Relay connector.

Some points:
- The connector supports both coordinated and uncoordinated modes, by a configuration toggle
- The connector has a pre-emptive cancel function (and config toggle) when using the coordinated version, as orders with < 30 seconds expiry aren't shown in the UI on Bamboo Relay
- Added in flight cancels list
- When placing limit orders its possible for the strategy to attempt to cancel them before the API has responded, so I added a in flight pending limit orders dictionary to address this
- Placing market orders now uses fillOrder or batchFillOrder calls on the smart contracts, as this is better for gas consumption
- Added an API failure timeout (3 seconds), to prevent the strategy from spamming the server endpoints on every tick
- Added support for the testnets to the connector (as needed this for testing anyway), but I had to modify a few other spots in the hummingbot code to do this (erc20_events_watcher.py/erc20_token.py)

For testing I've been running the integration tests on ropsten, and placing buy/sell DAI orders on the book (so something is sitting there).

Anyway this is a big PR, let me know what you think/I should change.

carlolm commented:
ok to test

chowryan commented on hummingbot/market/bamboo_relay/bamboo_relay_api_order_book_data_source.py:
```diff
@@ -37,47 +36,63 @@ class BambooRelayAPIOrderBookDataSource(OrderBookTrackerDataSource):
     @classmethod
-    def http_client(cls) -> aiohttp.ClientSession:
-        if cls._client is None:
+    def http_client(self) -> aiohttp.ClientSession:
```
our convention is to use cls for `@classmethod`

chowryan commented on hummingbot/market/bamboo_relay/bamboo_relay_api_order_book_data_source.py:
```diff
@@ -37,47 +36,63 @@ class BambooRelayAPIOrderBookDataSource(OrderBookTrackerDataSource):
 
     @classmethod
-    async def get_all_token_info(cls) -> Dict[str, any]:
+    async def get_all_token_info(self) -> Dict[str, any]:
```
our convention is to use cls for @classmethod

chowryan commented on hummingbot/market/bamboo_relay/bamboo_relay_api_order_book_data_source.py:
```diff
@@ -110,8 +125,14 @@ def order_book_class(self) -> BambooRelayOrderBook:
     @staticmethod
-    async def get_snapshot(client: aiohttp.ClientSession, trading_pair: str) -> Dict[str, any]:
-        async with client.get(f"{REST_BASE_URL}/markets/{trading_pair}/book") as response:
+    async def get_snapshot(client: aiohttp.ClientSession, 
```
static methods should not require self. pass in the api_prefix from the function that calls this function

chowryan commented on hummingbot/market/bamboo_relay/bamboo_relay_api_order_book_data_source.py:
```diff
@@ -134,7 +155,7 @@ def order_book_class(self) -> BambooRelayOrderBook:
             for trading_pair in trading_pairs:
                 try:
-                    snapshot: Dict[str, any] = await self.get_snapshot(client, trading_pair)
+                    snapshot: Dict[str, any] = await self.get_snapshot(client, trading_pair, self)
```
pass in api_prefix instead of self here

chowryan commented on hummingbot/market/bamboo_relay/bamboo_relay_market.pyx:
```diff
@@ -250,39 +282,54 @@ cdef class BambooRelayMarket(MarketBase):
     def __init__(self,
                  wallet: Web3Wallet,
                  ethereum_rpc_url: str,
+                 chain: EthereumChain = EthereumChain.MAIN_NET,
```
the initialized wallet instance knows which chain its on. check line 147 on web3_wallet_backend.py.

Arctek replied:
I exposed it on Web3Wallet as a property, given that makes the most sense?

carlolm commented:
This had a compile error:

```
Error compiling Cython file:
------------------------------------------------------------
...
        self._approval_tx_polling_task = None
        self._wallet = wallet
        self._use_coordinator = use_coordinator
        self._pre_emptive_soft_cancels = pre_emptive_soft_cancels
        self._latest_salt = -1
        if wallet.chain is EthereumChain.MAIN_NET:
                          ^
------------------------------------------------------------

hummingbot/market/bamboo_relay/bamboo_relay_market.pyx:345:27: undeclared name not builtin: EthereumChain
Traceback (most recent call last):
  File "/root/miniconda/envs/hummingbot/lib/python3.6/site-packages/Cython/Build/Dependencies.py", line 1244, in cythonize_one_helper
    return cythonize_one(*m)
  File "/root/miniconda/envs/hummingbot/lib/python3.6/site-packages/Cython/Build/Dependencies.py", line 1220, in cythonize_one
    raise CompileError(None, pyx_file)
Cython.Compiler.Errors.CompileError: hummingbot/market/bamboo_relay/bamboo_relay_market.pyx
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/root/miniconda/envs/hummingbot/lib/python3.6/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/root/miniconda/envs/hummingbot/lib/python3.6/multiprocessing/pool.py", line 44, in mapstar
    return list(map(*args))
  File "/root/miniconda/envs/hummingbot/lib/python3.6/site-packages/Cython/Build/Dependencies.py", line 1244, in cythonize_one_helper
    return cythonize_one(*m)
  File "/root/miniconda/envs/hummingbot/lib/python3.6/site-packages/Cython/Build/Dependencies.py", line 1220, in cythonize_one
    raise CompileError(None, pyx_file)
Cython.Compiler.Errors.CompileError: hummingbot/market/bamboo_relay/bamboo_relay_market.pyx
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "setup.py", line 143, in <module>
    main()
  File "setup.py", line 132, in main
    ], language="c++", language_level=3, nthreads=cpu_count),
  File "/root/miniconda/envs/hummingbot/lib/python3.6/site-packages/Cython/Build/Dependencies.py", line 1088, in cythonize
    result.get(99999)  # seconds
  File "/root/miniconda/envs/hummingbot/lib/python3.6/multiprocessing/pool.py", line 670, in get
    raise self._value
Cython.Compiler.Errors.CompileError: hummingbot/market/bamboo_relay/bamboo_relay_market.pyx
```

Arctek commented:
Ah sorry, I forget that commits to a PR will automatically re-run tests.
I'll fix it up locally first.

yzhang1994 commented on hummingbot/client/config/global_config_map.py:
```diff
@@ -99,6 +99,14 @@ def using_wallet() -> bool:
+                                                  prompt="Would you like to use the Bamboo Relay Coordinator >>> ",
+                                                  required_if=using_exchange("bamboo_relay"),
+                                                  is_secure=True),
+    "bamboo_relay_pre_emptive_soft_cancels":      ConfigVar(key="bamboo_relay_pre_emptive_soft_cancels",
```
This looks like a yes/no question, in which case it's good to add `type_str` = "bool" there. Also the value does not have to be marked as `is_secure` (what that does is it masks the string when user types it in, which is useful for API keys)

mifeng commented:
@Arctek I tested the new connector with a couple of strategies and run into some issues:

* pure market making: bot placed the initial bid and ask orders, but did not cancel/refresh them after 30 seconds.
* cross-exchage market making: bot placed the initial order but didn't adjust them. I used DAI-WETH on Bamboo vs ETH-USDT on Binance, which doesn't work since there's a mismatch between the base and quote assets. This caused errors in the logs.

Strategy config and log files:
[bamboo tests.zip](https://github.com/CoinAlpha/hummingbot/files/3379479/bamboo.tests.zip)

yzhang1994 commented:
@Arctek In order for your code to be packaged correctly, please review `setup.py` and include all relevant `abi.json` files in `package_data`

Arctek commented:
Hey Mike,

Thanks I'll take a look - For those logs it's complaining about "ONEUSDT" being unknown?
Also I tried this strategy out using DDEX instead and it's giving me similar issues with the pricing being messed up, i.e. -99%, 12816744% bid/ask.
So not sure where this pricing issue needs to be addressed? I assume it's going to happen across any other pair that might be inversely set across different exchanges.

In the case of the pure strategy - this is because radar and bamboo are hard coded to avoid cancelling orders - I've just added a check for this.

Also in your config files did you set `bamboo_relay_use_coordinator` to true? As otherwise it will be running in the previous, non-cancelling mode.

mifeng commented:
@Arctek I tested the new version - looks like there are still issues, possibly due to the fact that we've refactored the strategy and market classes during the current sprint.

I ran the pure MM and cross-exchange MM strategies and saw some issues - see below for screenshots, config files, and logs.

Regarding the DAI-WETH pair on Bamboo, you have DAI as the base asset (the denomination for order quantities) and WETH as the quote asset (the denomination for order prices). Other exchanges like DDEX, Radar, and Binance have this reversed - generally, I've found that stablecoins are used as the quote asset in many trading pairs.

[Archive 2.zip](https://github.com/CoinAlpha/hummingbot/files/3387327/Archive.2.zip)

Arctek commented:
@mifeng I've imported your config files from the zip - both of these initialize and run for me (see below for attached zip for screenshot/logs).
I made sure to pull, clean and recompile before testing.

When you tried did you make sure to rebuild the project, i.e. ./clean && ./compile
Because the error in the `pure mm.log` file you sent me:
`TypeError: __init__() got an unexpected keyword argument 'use_coordinator'`
Would only occur if the .pxy file hadn't been rebuilt - because it only exists as an addition with this PR.

With the DAI-WETH pair I can see your point, what I might do is add support for it both ways in the API on my side - this will be a quicker short term fix, as I can't easily reverse the book as it is live now.

[test.zip](https://github.com/CoinAlpha/hummingbot/files/3387417/test.zip)

mifeng commented:
Thanks @Arctek - got it working now after I ran `./clean`. So far so good - I've been running it for the past few hours and it's working.

FYI since we have already code freezed for next Monday's release, I think we should target getting this into `master` in the following release.
* Did you address @yzhang1994's comments above in the most recent commits?
* Can you add a page to `/documentation/connectors/` folder similar to https://docs.hummingbot.io/connectors/radar-relay/? This page is intended to provide the basic info regarding exchange specifics that users may need. In particular: min order sizes, fees (if any), and anything else users may need to know. For instance, perhaps a warning about how the `DAI-WETH` pair is currently incompatible with `WETH-DAI` pair on Radar/DDEX.
* Add `(y/n)` to the question `Would you like to use the Bamboo Relay Coordinator >>>` and the other one. We will add these prompts for other boolean questions as well.

Arctek commented:
Hey, that's good news!

Sure that's fine regarding the next release target, 

- Yes I made the changes in the above commits that @yzhang1994 suggested, so these should be covered (not sure why github didn't mark these as out of date)
- I'll see about writing up the docs this week for settings and the points you've mentioned.
- Both of those questions should be marked as y/n in the latest commit. Though I think if the default configuration contains these then it doesn't prompt for them?

Arctek commented:
Alright

- Added a documentation page, let me know if this reads ok
- Added (y/n) to the Bamboo specific questions and set these to be null by default so it prompts users the first time
- I've applied the InFlightOrder refactoring to Bamboo Relay, and some other pieces that had been applied against the other connectors

I also fixed the bug in the strategy_base.pxy, where it wasn't passing kwargs to the c_buy/c_sell functions (which is needed for the expiration_ts value).
I had to change this to kwargs (was kargs) in the bamboo/radar connectors too in order for this to be correct.

mifeng commented:
I tested the PR after the most recent changes:

1. I see the following error both in the latest commits and in commit `b656a99` (before the merge of `development` branch):
```
2019-07-22 10:50:08,653 - hummingbot.client.hummingbot_application - ERROR - 'is_coordinated'
Traceback (most recent call last):
  File "/Users/feng/Code/hummingbot/hummingbot/client/command/start_command.py", line 91, in start_market_making
    self.markets_recorder.restore_market_states(config_path, market)
  File "/Users/feng/Code/hummingbot/hummingbot/market/markets_recorder.py", line 140, in restore_market_states
    market.restore_tracking_states(market_states.saved_state)
  File "hummingbot/market/bamboo_relay/bamboo_relay_market.pyx", line 294, in hummingbot.market.bamboo_relay.bamboo_relay_market.BambooRelayMarket.restore_tracking_states
    key: BambooRelayInFlightOrder.from_json(value)
  File "hummingbot/market/bamboo_relay/bamboo_relay_in_flight_order.pyx", line 119, in hummingbot.market.bamboo_relay.bamboo_relay_in_flight_order.BambooRelayInFlightOrder.from_json
    is_coordinated=bool(data["is_coordinated"]),
KeyError: 'is_coordinated'
```

2. The documentation pages looks good. In `/documentation/mkdocs.yml`, add an entry under Connectors for Bamboo Relay. Also, please revise the Bamboo link in `/documentation/docs/connectors/index.md to point to the docs page.

Arctek commented:
Hey Mike,

1. This is because of the tracking states json that's stored in the database - as that field is new. i.e. if you delete the database under the data dir it will run (hummingbot_trades.sqlite).
When users do a hummingbot upgrade is this database re-created?
Or should I add error handling in the order restoration code, i.e. from_json?

2. Just added this in the docs.

Arctek commented:
Ok - it should work now with that older database gracefully and just convert missing/changed fields. 

In future there will probably need to be SQL migrations to update saved InFlightOrders and add/update fields, to prevent it becoming too messy.

mifeng commented:
Tested all 4 strategies and it appears to work well. I've approved but I think someone from engineering should review the code before we merge.
